atlp-rwanda/Team-furebo-E-commerce-bn
# Ft: two factor authentication #184651335

#### What does this PR do?
This PR introduces a new feature that allows Sellers to enable two-factor authentication (2FA) on their accounts for added security. With 2FA enabled, authentication to the application will require an extra factor which is an authentication code in addition to the regular login credentials.

#### Description of Task to be completed?
- Added a new column "enable2FA" to the Users table to indicate whether 2FA is enabled for a particular user.
- Added a new column "twoFactorAuthKey" to the Users table to hold the generate secret key at each login when 2FA is enabled
- Implemented an endpoint '/api/2fa/enable2faForMerchant' that requests a jwt token to turn on 2FA
- Implemented an endpoint '/api/2fa/disable2faForMerchant' that requests a jwt token to turn off 2FA
- Implemented an endpoint '/api/2fa/resendOTP' that requests a jwt token to resend authentication code in case it expires
- Implemented a new endpoint "/api/2fa/verify" that requests for a 2FA authentication code upon login when 2FA is enabled for the particular user.
-The 2FA token is validated using the speakeasy package and if valid, the user is allowed to log in

#### How should this be manually tested?

 Using a tool like postman or swagger:

1. Navigate to the http://127.0.0.1:5002/api/login post request with valid credentials
2. Use the token provide at login and set authorization header at this '/api/2fa/enable2faForMerchant' endpoint
3. Once 2fa is enabled, login again and authentication code will be sent to the user email.
4. Use the authentication code received on the user's registered email, give this "/api/2fa/verify"  a parameter "code" with the authentication code and set  the authorization header with the jwt token provided at login.
5.  If the authentication code has expired, navigate to this  '/api/2fa/resendOTP' endpoint and set it's authorization header with the toke from login
6.  Navigate back to the verification endpoint and use the newly generated code

#### What are the relevant pivotal tracker/Trello stories?
184651335
